solr/core/src/java/org/apache/solr/cluster/placement/impl/AttributeFetcherImpl.java

```diff
@@ -113,72 +90,46 @@ public AttributeFetcher fetchFrom(Set<Node> nodes) {
-          nodeToDiskType.put(node, DiskHardwareType.SSD);
-        }
-        // unknown disk type: insert no value, returned optional will be empty
-      });
-    }
-    if (requestedNodeFreeDisk) {
-      allSnitchTagsToInsertion.put(SolrClientNodeStateProvider.Variable.FREEDISK.tagName,
-          // Convert from bytes to GB
-          (node, value) -> nodeToFreeDisk.put(node, ((Number) value).longValue() / 1024 / 1024 / 1024));
-    }
-    if (requestedNodeTotalDisk) {
-      allSnitchTagsToInsertion.put(SolrClientNodeStateProvider.Variable.TOTALDISK.tagName,
-          // Convert from bytes to GB
-          (node, value) -> nodeToTotalDisk.put(node, ((Number) value).longValue() / 1024 / 1024 / 1024));
-    }
-    if (requestedNodeHeapUsage) {
-      allSnitchTagsToInsertion.put(ImplicitSnitch.HEAPUSAGE,
-          (node, value) -> nodeToHeapUsage.put(node, ((Number) value).doubleValue()));
-    }
-    if (requestedNodeSystemLoadAverage) {
-      allSnitchTagsToInsertion.put(ImplicitSnitch.SYSLOADAVG,
-          (node, value) -> nodeToSystemLoadAverage.put(node, ((Number) value).doubleValue()));
-    }
-    for (String sysPropSnitch : requestedNodeSystemPropertiesSnitchTags) {
-      final Map<Node, String> sysPropMap = Maps.newHashMap();
-      syspropSnitchToNodeToValue.put(sysPropSnitch, sysPropMap);
+    Map<String, BiConsumer<Node, Object>> allSnitchTagsToInsertion = new HashMap<>();
+    for (String sysPropSnitch : requestedNodeSystemSnitchTags) {
+      final Map<Node, String> sysPropMap = new HashMap<>();
+      systemSnitchToNodeToValue.put(sysPropSnitch, sysPropMap);
       allSnitchTagsToInsertion.put(sysPropSnitch, (node, value) -> sysPropMap.put(node, (String) value));
     }
-    for (String metricSnitch : requestedNodeMetricSnitchTags) {
-      final Map<Node, Double> metricMap = Maps.newHashMap();
-      metricSnitchToNodeToValue.put(metricSnitch, metricMap);
-      allSnitchTagsToInsertion.put(metricSnitch, (node, value) -> metricMap.put(node, (Double) value));
+    for (NodeMetric<?> metric : requestedNodeMetricSnitchTags) {
+      final Map<Node, Object> metricMap = new HashMap<>();
+      metricSnitchToNodeToValue.put(metric, metricMap);
+      String metricSnitch = getMetricSnitchTag(metric);
+      allSnitchTagsToInsertion.put(metricSnitch, (node, value) -> metricMap.put(node, metric.convert(value)));
     }
+    requestedCollectionMetrics.forEach((collection, tags) -> {
+      Set<String> collectionTags = tags.stream().map(tag -> tag.getInternalName()).collect(Collectors.toSet());
+      collection.shards().forEach(shard ->
+          shard.replicas().forEach(replica -> {
+            Set<String> perNodeInternalTags = nodeToReplicaInternalTags
+                .computeIfAbsent(replica.getNode(), n -> new HashSet<>());
+            perNodeInternalTags.addAll(collectionTags);
+          }));
+    });
 
     // Now that we know everything we need to fetch (and where to put it), just do it.
     for (Node node : nodes) {
```

Comment:
Maybe add a comment here saying fetching could be made concurrent from all nodes at once for perf improvement?